File: 3147-taking-maximum-energy-from-the-mystic-dungeon/3147-taking-maximum-energy-from-the-mystic-dungeon.cpp
class Solution {
public:
    int maximumEnergy(vector<int>& energy, int k) {
        int n = energy.size();
        int end = n - 1;
        long long ans = INT_MIN;
        while (end >= n - k) {
            long long temp = 0;
            for (int i = end; i >= 0; i -= k) {
                temp += (long long)energy[i];
                ans = max(ans, temp);
            }
            end--;
        }
        return ans;
    }
};

File: 3147-taking-maximum-energy-from-the-mystic-dungeon/3147-taking-maximum-energy-from-the-mystic-dungeon.py
class Solution:
    def maximumEnergy(self, energy: List[int], k: int) -> int:
        n = len(energy)
        dp = [-1] * n
        ans = float("-inf")
        for i in range(n - 1, -1, -1):
            if i + k < n:
                dp[i] = dp[i + k] + energy[i]
            else:
                dp[i] = energy[i]
            ans = max(ans, dp[i])
        return ans
